x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/logging/Throttler.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.logging;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.Closeable;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.inference.InferencePlugin.UTILITY_THREAD_POOL_NAME;
+
+/**
+ * A class that throttles calls to a logger. If a log call is made during the throttle period a counter is incremented.
+ * If a log call occurs after the throttle period, then the call will proceed, and it will include a message like
+ * "repeated X times" to indicate how often the message was attempting to be logged.
+ */
+public class Throttler implements Closeable {
```

Comment:
I refactored this class to use a `ConcurrentHashMap` to handle counting the log message occurrences. This class now always allows the first call. If calls are made before a duration completes, it just increments a counter.

To avoid having the map grow indefinitely I set a reoccurring task to run every day (it's configurable too) to clear the map. We could lose some log messages because of that, but I think that's ok.